Use type-based ranges for DAC outputs

mullr opened the pull request:
- Add the generic ranges::Bounded wrapper type, built upon typenum
- Replace u16 parameters to the dac module with Bounded<u16, 0, 4095>,
  effectively a compile time check that the appropriate range checks or clamping
  has been done
- Implement Bounded-based clamping at all call sites

mullr committed:
Use type-based ranges for DAC outputs

- Add the generic ranges::Bounded wrapper type, built upon typenum
- Replace u16 parameters to the dac module with Bounded<u16, 0, 4095>,
  effectively a compile time check that the appropriate range checks or clamping
  has been done
- Implement Bounded-based clamping at all call sites

ZackPierce commented on src/ranges.rs:
```diff
@@ -0,0 +1,77 @@
+
+#[derive(Debug)]
+pub struct Bounded<T,L,U> {
+    pub val: T,
```
Doesn't the fact that this field is `pub` mean that it can be directly mutated to any value of type `T` by an external actor with mutable access to the `Bounded` instance?  Seems like a read-only getter (and likely a constructor) would be appropriate here to prevent that possibility.

mullr replied:
oh good grief, yes of course that's true. I'll add two setters, one checking and the other proof-carrying.

mullr replied:
Or maybe I can just leave out the setters for now, until they're needed. At the moment all ranges are constructed via the `clamp` function. (I have a constructor, I just took it out to keep this minimal for now)

ZackPierce replied:
I think something like a constructor and all-private-fields would be necessary for me to feel good about using this thing with the level of confidence I was expecting -- namely, I want to be able to use instances without ever worrying about if they are in an invalid state (and instead just worry about getting things transformed into that type)

mullr replied:
Ok. I've added a commit to make the value private. W.r.t. a constructor: I believe `clamp` is such a function. I have some code in another repo that provides `new(val: T) -> Result<Bounded<T, L, U>, ...>`, but it wasn't being used here so I deleted it from this version.

ZackPierce reviewed (changes requested):
One API question/nit, otherwise seems sane and cool.

mullr committed:
Make Bounded's val field private

This means nobody can mutate it from the outside. Add a val() getter so the
value can be observed.

ZackPierce reviewed (approved):
Looks good!

jonlamb-gh reviewed (approved):
Very cool!
